engine/runtime-compiler/src/main/java/org/enso/compiler/phase/ImportResolverForIR.java

```diff
@@ -0,0 +1,163 @@
+    return e.qualifiedName().item();
+  }
+
+  @Override
+  protected final java.util.List<Export.Module> exportsFor(Module module, String impName) {
+    java.util.List<Export.Module> exp = CollectionConverters.SeqHasAsJava(module.exports()).asJava().stream().map(e -> switch (e) {
+      case Export.Module ex when ex.name().name().equals(impName) -> ex;
+      case null, default -> null;
+    }).filter(Objects::nonNull).toList();
+    return exp;
+  }
+
+  @Override
+  protected final boolean isAll(Export.Module ex) {
+    return ex.isAll();
+  }
+
+  @Override
+  protected final java.util.List<String> onlyNames(Export.Module ex) {
+    if (ex.onlyNames().isEmpty()) {
+      return null;
+    }
+    var list = CollectionConverters.SeqHasAsJava(ex.onlyNames().get().map(n -> n.name())).asJava();
+    return list;
+  }
+
+  @Override
+  protected final java.util.List<String> hiddenNames(Export.Module ex) {
+    if (ex.hiddenNames().isEmpty()) {
+      return null;
+    }
+    var list = CollectionConverters.SeqHasAsJava(ex.hiddenNames().get().map(n -> n.name())).asJava();
+    return list;
+  }
+
+  @Override
+  protected final java.util.List<BindingsMap.ResolvedType> definedEntities(Import.Module name) {
+    var parts = partsForImport(name);
+    var last = parts.size() - 1;
+    var modName = String.join(".", parts.subList(0, last).stream().toList());
+    var compiler = this.getCompiler();
+    var optionMod = compiler.getModule(modName);
+    if (optionMod.isEmpty()) {
+      return null;
+    }
+    var mod = optionMod.get();
+    compiler.ensureParsed(mod);
+    var b = mod.getBindingsMap();
+    if (b == null) {
+      compiler.context().updateModule(mod, u -> {
+        u.invalidateCache();
+        u.ir(null);
+        u.compilationStage(CompilationStage.INITIAL);
+      });
+      compiler.ensureParsed(mod, false);
+      b = mod.getBindingsMap();
+    }
+    var entitiesStream = b.definedEntities().map(e -> switch (e) {
+      case BindingsMap.Type t -> {
+        assert e.name().equals(t.name()) : e.name() + " != " + t.name();
+        var res = new ResolvedType(new BindingsMap$ModuleReference$Concrete(mod), t);
+        assert e.name().equals(res.tp().name()) : e.name() + " != " + res.tp().name();
+        yield res;
+      }
+      case null, default -> null;
+    }).filter(Objects::nonNull);
+    var entities = CollectionConverters.SeqHasAsJava(entitiesStream).asJava();
+    return entities;
+  }
+
+  @Override
+  protected final CompilerContext.Module loadLibraryModule(LibraryName libraryName, String moduleName) throws IOException {
+    var compiler = this.getCompiler();
+    var repo = compiler.packageRepository();
+    var foundLib = repo.ensurePackageIsLoaded(libraryName);
+    if (foundLib.isRight()) {
+      var moduleOption = compiler.getModule(moduleName);
+      if (moduleOption.isDefined()) {
+        return moduleOption.get();
+      } else {
+        return null;
+      }
```

Comment:
I don't think this is correct. `getOrElse` needs a function returning a value. E.g. it would be `getOrElse(() -> null)` - but I don't want to create lambda functions.